### PR TITLE
Add Invocations to Object Oriented API

### LINKED ISSUE #317: Workflow.run() doesn't wait and doesn't return outputs
https://github.com/galaxyproject/bioblend/blob/c1a13cd7a3fa749754d0f7990bafcbfd1ee5421f/bioblend/galaxy/objects/wrappers.py#L460-L467

The api returns an empty 'outputs' dict in the invocation information. This causes this code to immediately exit the loop and produce an empty list.

The code needs to wait for the 'outputs' dict to become populated by Galaxy.

### LINKED ISSUE #391: Add an `invoke()` method to the BioBlend.objects `Workflow` class
> Doesn't need to be part of this PR, but the next step to close the circle would be to add an `invoke()` method to the BioBlend.objects `Workflow` class, which would return an `Invocation` object (and use it here).
> 
> _Originally posted by @nsoranzo in https://github.com/galaxyproject/bioblend/pull/387#discussion_r620500646_

### PR DESCRIPTION
**Pull request by rikeshi**
This PR adds the following: 

- Invocation class to galaxy/objects/wrappers.py, with two helper classes InvocationStep and InvocationPreview
- ObjInvocationClient in galaxy/objects/client.py, accessible as invocations in the objects GalaxyInstance
- Tests for the Invocation and ObjInvocationClient methods, located in tests/TestGalaxyObjects.py

Close #317, close #391

### CODE REVIEW

File: bioblend/galaxy/objects/client.py [hunk 1]
```diff
@@ -323,6 +323,53 @@ def delete(self, id_=None, name=None):
...
+        return self.gi.invocations.get_invocation_step_jobs_summary(step_id)
+
+    def report(self, id: str) -> dict:
+        return self.gi.invocations.get_invocation_report(id)
+
+    def save_report_pdf(self, id: str, file_path: str, chunk_size: int = bioblend.CHUNK_SIZE):
+        self.gi.invocations.get_invocation_report_pdf(id, file_path, chunk_size)
+
+    def biocompute_object(self, id: str) -> dict:
+        return self.gi.invocations.get_invocation_biocompute_object(id)
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/client.py:370]: I think all these methods should be moved to the `Invocation` class below.

**Review by nsoranzo — COMMENTED** [review]
Thanks a lot for starting this!

File: bioblend/galaxy/objects/client.py [hunk 2]
```diff
@@ -323,6 +323,55 @@ def delete(self, id_=None, name=None):
...
+
+    def list(self, include_terminal=True):
+        """
+        Get full listing of invocations.
+        This can be slow!
+
+        :param include_terminal: bool
+        :param: whether to include invocations in terminal states
+
+        :rtype: list of Invocation
```
Comment (simonbray) [inline bioblend/galaxy/objects/client.py:364]:
```suggestion
        :rtype: list of Invocations
```

Comment (simonbray) [reply bioblend/galaxy/objects/client.py:364]: Can / should we add filtering here?

File: bioblend/galaxy/objects/wrappers.py [hunk 3]
```diff
@@ -492,6 +515,165 @@ def delete(self):
...
+        Get a summary for this invocation's step jobs.
+
+        :rtype: list of dicts
+        :param: step job summaries
+        """
+        return self.gi.gi.invocations.get_invocation_step_jobs_summary(self.id)
+
+    def report(self):
+        """
+        Get a report for this invocation.
```
Comment (simonbray) [inline bioblend/galaxy/objects/wrappers.py:638]: what kind: markdown, json...?

File: bioblend/galaxy/objects/wrappers.py [hunk 4]
```diff
@@ -492,6 +515,165 @@ def delete(self):
...
+
+        :rtype: dict
+        :param: invocation report
+        """
+        return self.gi.gi.invocations.get_invocation_report(self.id)
+
+    def save_report_pdf(self, file_path, chunk_size=bioblend.CHUNK_SIZE):
```
Comment (simonbray) [inline bioblend/galaxy/objects/wrappers.py:645]: Maybe add a docstring here?

File: bioblend/_tests/TestGalaxyObjects.py [hunk 5]
```diff
@@ -234,6 +277,207 @@ def setUp(self):
...
+        report = inv.report()
+        assert report['workflows'] == {self.workflow_id: {'name': 'paste_columns'}}
+
+    def test_save_report_pdf(self):
+        inv = self._obj_invoke_workflow()
+        try:
+            inv.save_report_pdf('report.pdf')
+        except Exception:
+            # This can fail if dependencies as weasyprint are not installed on the Galaxy server
+            pass
```
Comment (simonbray) [inline bioblend/_tests/TestGalaxyObjects.py:406]: This test is doing nothing IMO - maybe just remove?

**Review by simonbray — COMMENTED** [review]
Looks mostly ok to me

**Review by simonbray — APPROVED** [review]

File: bioblend/galaxy/objects/wrappers.py [hunk 6]
```diff
@@ -177,6 +177,29 @@ def gi_module(self):
         return self.gi.workflows
 
 
+class InvocationStep(Wrapper):
+    """
+    Abstract base class for invocation steps.
+    """
+    BASE_ATTRS = Wrapper.BASE_ATTRS + (
+        'job_id',
+        'action',
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/wrappers.py:186]:
```suggestion
        'action',
        'job_id',
```

File: bioblend/galaxy/objects/wrappers.py [hunk 7]
```diff
@@ -490,6 +513,174 @@ def delete(self):
...
+    This causes the steps of a workflow to be executed in sequential order.
+    """
+    BASE_ATTRS = Wrapper.BASE_ATTRS + (
+        'workflow_id',
+        'history_id',
+        'inputs',
+        'state',
+        'steps',
+        'update_time',
+        'uuid',
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/wrappers.py:528]:
```suggestion
        'history_id',
        'inputs',
        'state',
        'steps',
        'update_time',
        'uuid',
        'workflow_id',
```

File: bioblend/galaxy/objects/wrappers.py [hunk 8]
```diff
@@ -490,6 +513,174 @@ def delete(self):
...
+        """
+        Cancel this invocation.
+
+        .. note::
+          On success, this method updates the Invocation object's internal variables.
+        """
+        inv_dict = self.gi.gi.invocations.cancel_invocation(self.id)
+        self.__init__(inv_dict, gi=self.gi)
+
+    def update(self):
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/wrappers.py:593]:
```suggestion
    def refresh(self):
```

File: bioblend/galaxy/objects/client.py [hunk 9]
```diff
@@ -323,6 +323,74 @@ def delete(self, id_=None, name=None):
...
+        :type id: str
+        :param id: invocation ID
+
+        :rtype: Invocation
+        :param: invocation object
+        """
+        inv_dict = self.gi.invocations.show_invocation(id)
+        return wrappers.Invocation(inv_dict, self.obj_gi)
+
+    def get_previews(self):
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/client.py:346]:
```suggestion
    def get_previews(self) -> List[wrappers.InvocationPreview]:
```

File: bioblend/galaxy/objects/client.py [hunk 10]
```diff
@@ -323,6 +323,74 @@ def delete(self, id_=None, name=None):
...
+        Get previews of all invocations.
+
+        :rtype: list of InvocationPreview
+        :param: previews of invocations
+        """
+        inv_list = self.gi.invocations.get_invocations()
+        return [wrappers.InvocationPreview(inv_dict, self.obj_gi) for inv_dict in inv_list]
+
+    def list(self, workflow_id=None, history_id=None, user_id=None,
+             include_terminal=True, limit=None):
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/client.py:357]:
```suggestion
    def list(
    	self,
    	workflow_id=None,
    	history_id=None,
    	user_id=None,
        include_terminal=True,
        limit=None
    ) -> List[wrappers.Invocation]:
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 11]
```diff
@@ -234,6 +277,203 @@ def setUp(self):
...
+        inv = self._obj_invoke_workflow()
+        inv.cancel()
+        self.assertEqual(inv.state, 'cancelled')
+
+    def test_wait(self):
+        inv = self._obj_invoke_workflow()
+        inv.wait()
+        self.assertEqual(inv.state, 'scheduled')
+
+    def test_update(self):
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:357]:
```suggestion

    def test_refresh(self):
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 12]
```diff
@@ -234,6 +277,203 @@ def setUp(self):
...
+        inv = self._obj_invoke_workflow()
+        # use wait_for_invocation() directly, because inv.wait() will update inv automatically
+        self.gi.gi.invocations.wait_for_invocation(inv.id)
+        self.assertEqual(inv.state, 'new')
+        inv.update()
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:362]:
```suggestion
        inv.refresh()
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 13]
```diff
@@ -234,6 +277,203 @@ def setUp(self):
...
+        invocation_id = self.gi.gi.workflows.invoke_workflow(
+            self.workflow_pause_id,
+            inputs={"0": {"src": "hda", "id": self.dataset_id}},
+            history_id=self.history_id,
+        )['id']
+        inv = self.gi.invocations.get(invocation_id)
+        for _ in range(20):
+            with self.assertRaises(bioblend.TimeoutException):
+                inv.wait(maxwait=0.5, interval=0.5)
+            inv.update()
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:375]:
```suggestion
            inv.refresh()
```

**Review by nsoranzo — COMMENTED** [review]
Thanks, looking great, just a couple of suggestions.

I have [a commit](https://github.com/nsoranzo/bioblend/commit/4e2123ad5f5f27fbb762c17e5e335236b85a692a) in my fork which refactors and simplify the BioBlend.objects classes.
It's ready to push upstream, but then you would need to apply similar changes to this PR. Would you prefer to get this in first?

**Comment by rikeshi** [comment]
I don't mind. If your changes are ready then commit them first and I can update the PR.

**Comment by nsoranzo** [comment]
> I don't mind. If your changes are ready then push them first and I can update the PR.

OK, thanks @rikeshi ! I've pushed the commit, let me know if you need help in fixing the conflicts or updating the methods.

File: bioblend/galaxy/objects/client.py [hunk 14]
```diff
@@ -326,6 +327,74 @@ def delete(self, id_=None, name=None):
...
+        :param: previews of invocations
+        """
+        inv_list = self.gi.invocations.get_invocations()
+        return [wrappers.InvocationPreview(inv_dict, self.obj_gi) for inv_dict in inv_list]
+
+    def list(
+        self,
+        workflow_id=None,
+        history_id=None,
+        user_id=None,
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/client.py:358]: In general BioBlend.objects methods shouldn't be passed ids (apart from `get()`), so I think these parameters should be of type `wrappers.Workflow` and `wrappers.History`, dropping `user_id` for now (the objects interface is not tailored towards admins any way).

**Review by nsoranzo — COMMENTED** [review]
Thanks, this looks great!

File: bioblend/_tests/TestGalaxyObjects.py [hunk 15]
```diff
@@ -232,6 +275,203 @@ def setUp(self):
...
+    @classmethod
+    def setUpClass(cls):
+        super().setUp(cls)
+        path = test_util.get_abspath(os.path.join('data', 'paste_columns.ga'))
+        self.workflow_id = cls.gi.gi.workflows.import_workflow_from_local_path(path)['id']
+        cls.history_id = cls.gi.gi.histories.create_history(name="TestGalaxyInvocations")["id"]
+        dataset_id = cls.gi.gi.tools.paste_content('1\t2\t3', cls.history_id)["outputs"][0]["id"]
+        dataset = {'src': 'hda', 'id': dataset_id}
+        self.invocation_id = cls.gi.gi.workflows.invoke_workflow(
+            self.workflow_id,
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py]:
```suggestion
        cls.invocation_id = cls.gi.gi.workflows.invoke_workflow(
            cls.workflow_id,
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 16]
```diff
@@ -232,6 +275,203 @@ def setUp(self):
...
+
+
+@test_util.skip_unless_galaxy('release_19.09')
+class TestObjInvocationClient(GalaxyObjectsTestBase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUp(cls)
+        path = test_util.get_abspath(os.path.join('data', 'paste_columns.ga'))
+        self.workflow_id = cls.gi.gi.workflows.import_workflow_from_local_path(path)['id']
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py]:
```suggestion
        cls.workflow_id = cls.gi.gi.workflows.import_workflow_from_local_path(path)['id']
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 17]
```diff
@@ -232,6 +275,203 @@ def setUp(self):
...
+        cls.history_id = cls.gi.gi.histories.create_history(name="TestGalaxyInvocations")["id"]
+        dataset_id = cls.gi.gi.tools.paste_content('1\t2\t3', cls.history_id)["outputs"][0]["id"]
+        dataset = {'src': 'hda', 'id': dataset_id}
+        self.invocation_id = cls.gi.gi.workflows.invoke_workflow(
+            self.workflow_id,
+            inputs={'Input 1': dataset, 'Input 2': dataset},
+            history_id=cls.history_id,
+            inputs_by='name'
+        )['id']
+        cls.inv = cls.gi.invocations.get(invocation_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py]:
```suggestion
        cls.inv = cls.gi.invocations.get(cls.invocation_id)
```

File: bioblend/galaxy/objects/client.py [hunk 18]
```diff
@@ -326,6 +327,70 @@ def delete(self, id_=None, name=None):
...
+        """
+        Get full listing of workflow invocations, or select a subset
+        by specifying optional arguments for filtering (e.g. a workflow).
+
+        :type workflow: wrappers.Workflow
+        :param workflow: Include only invocations associated with
+          this workflow
+
+        :type history: str
+        :param history_id: Include only invocations associated with
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/client.py:370]:
```suggestion
        :param history: Include only invocations associated with
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 19]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+    @classmethod
+    def setUpClass(cls):
+        super().setUp(cls)
+        cls.inv = wrappers.Invocation(SAMPLE_INV_DICT)
+        path = test_util.get_abspath(os.path.join('data', 'paste_columns.ga'))
+        path_pause = test_util.get_abspath(os.path.join('data', 'test_workflow_pause.ga'))
+        workflow_id = cls.gi.gi.workflows.import_workflow_from_local_path(path)['id']
+        workflow_pause_id = cls.gi.gi.workflows.import_workflow_from_local_path(path_pause)['id']
+        cls.workflow = cls.gi.workflows.get(workflow_id)
+        cls.workflow_pause = cls.gi.workflows.get(workflow_pause_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:290]:
```suggestion
        with open(SAMPLE_FN) as f:
            cls.workflow = self.gi.workflows.import_new(f.read())
        path_pause = test_util.get_abspath(os.path.join('data', 'test_workflow_pause.ga'))
        with open(path_pause) as f:
            cls.workflow_pause = self.gi.workflows.import_new(f.read())
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 20]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+        cls.history = cls.gi.histories.create(name="TestInvocation")
+        cls.dataset_id = cls.gi.gi.tools.paste_content('1\t2\t3', cls.history.id)["outputs"][0]["id"]
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:292]:
```suggestion
        cls.dataset_id = cls.history.paste_content('1\t2\t3').id
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 21]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+@test_util.skip_unless_galaxy('release_19.09')
+class TestObjInvocationClient(GalaxyObjectsTestBase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUp(cls)
+        path = test_util.get_abspath(os.path.join('data', 'paste_columns.ga'))
+        cls.workflow_id = cls.gi.gi.workflows.import_workflow_from_local_path(path)['id']
+        cls.history_id = cls.gi.gi.histories.create_history(name="TestGalaxyObjInvocationClient")["id"]
+        dataset_id = cls.gi.gi.tools.paste_content('1\t2\t3', cls.history_id)["outputs"][0]["id"]
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:423]:
```suggestion
        with open(SAMPLE_FN) as f:
            cls.workflow = self.gi.workflows.import_new(f.read())
        cls.history = cls.gi.histories.create(name="TestGalaxyObjInvocationClient")
        dataset_id = cls.history.paste_content('1\t2\t3').id
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 22]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+        dataset = {'src': 'hda', 'id': dataset_id}
+        cls.invocation_id = cls.gi.gi.workflows.invoke_workflow(
+            cls.workflow_id,
+            inputs={'Input 1': dataset, 'Input 2': dataset},
+            history_id=cls.history_id,
+            inputs_by='name'
+        )['id']
+        cls.inv = cls.gi.invocations.get(cls.invocation_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:431]:
```suggestion
        cls.inv = self.workflow.invoke(
            inputs={'Input 1': dataset, 'Input 2': dataset},
            history=cls.history,
            inputs_by='name',
        )
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 23]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+        self.assertEqual(inv_preview.state, 'scheduled')
+        self.assertEqual(inv_preview.update_time, self.inv.update_time)
+        self.assertEqual(inv_preview.uuid, self.inv.uuid)
+
+    def test_list(self):
+        invs = self.gi.invocations.list()
+        inv = next(i for i in invs if i.id == self.invocation_id)
+        self.assertEqual(inv.id, self.invocation_id)
+        self.assertEqual(inv.workflow_id, self.workflow_id)
+        self.assertEqual(inv.history_id, self.history_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:463]:
```suggestion
        inv = next(i for i in invs if i.id == self.inv.id)
        self.assertEqual(inv.id, self.inv.id)
        self.assertEqual(inv.workflow_id, self.workflow.id)
        self.assertEqual(inv.history_id, self.history.id)
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 24]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+        self.assertEqual(inv.update_time, self.inv.update_time)
+        self.assertEqual(inv.uuid, self.inv.uuid)
+
+    def test_get_previews(self):
+        previews = self.gi.invocations.get_previews()
+        self.assertSetEqual({type(preview) for preview in previews}, {wrappers.InvocationPreview})
+        inv_preview = next(p for p in previews if p.id == self.invocation_id)
+        self.assertEqual(inv_preview.id, self.invocation_id)
+        self.assertEqual(inv_preview.workflow_id, self.workflow_id)
+        self.assertEqual(inv_preview.history_id, self.history_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:453]:
```suggestion
        inv_preview = next(p for p in previews if p.id == self.inv.id)
        self.assertEqual(inv_preview.id, self.inv.id)
        self.assertEqual(inv_preview.workflow_id, self.workflow.id)
        self.assertEqual(inv_preview.history_id, self.history.id)
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 25]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.gi.gi.histories.delete_history(cls.history_id, purge=True)
+
+    def test_get(self):
+        inv = self.gi.invocations.get(self.inv.id)
+        self.assertEqual(inv.id, self.invocation_id)
+        self.assertEqual(inv.workflow_id, self.workflow_id)
+        self.assertEqual(inv.history_id, self.history_id)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:442]:
```suggestion
        self.assertEqual(inv.id, self.inv.id)
        self.assertEqual(inv.workflow_id, self.workflow.id)
        self.assertEqual(inv.history_id, self.history.id)
```

File: bioblend/_tests/TestGalaxyObjects.py [hunk 26]
```diff
@@ -232,6 +275,201 @@ def setUp(self):
...
+            inputs={'Input 1': dataset, 'Input 2': dataset},
+            history_id=cls.history_id,
+            inputs_by='name'
+        )['id']
+        cls.inv = cls.gi.invocations.get(cls.invocation_id)
+        cls.inv.wait()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.gi.gi.histories.delete_history(cls.history_id, purge=True)
```
Comment (nsoranzo) [inline bioblend/_tests/TestGalaxyObjects.py:436]:
```suggestion
        cls.history.delete(purge=True)
```

File: bioblend/galaxy/objects/wrappers.py [hunk 27]
```diff
@@ -479,6 +495,332 @@ def delete(self):
...
+                       Galaxy API. Alternatively workflow steps may be addressed by
+                       the label that can be set in the workflow editor. If using
+                       uuid or label you need to also set the ``inputs_by`` parameter
+                       to ``step_uuid`` or ``name``.
+
+        :type params: dict
+        :param params: A mapping of non-datasets tool parameters (see below)
+
+        :type history: str
+        :param history_id: The history in which to store the workflow
```
Comment (nsoranzo) [inline bioblend/galaxy/objects/wrappers.py:527]:
```suggestion
        :param history: The history in which to store the workflow
```

**Comment by nsoranzo** [comment]
Thanks @rikeshi for adding the `invoke()` method!

Can you please add a deprecation note to `Workflow.run()`:
```python
        .. deprecated:: 0.16.0
           Use :meth:`invoke` instead.
```

Then it will be ready to merge!

**Comment by simonbray** [comment]
Thanks a lot @rikeshi!

**Comment by rikeshi** [comment]
Thanks for reviewing @simonbray and @nsoranzo !

**Pull request merged**